arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/popup/PopupViewModel.kt

```diff
@@ -0,0 +1,266 @@
+    val popup: LiveData<Popup> = _popup
+
+    private val _popupManager = MutableLiveData<PopupManager>()
+    // The manager for the Popup responsible for viewing and editing of the Popup
+    val popupManager: LiveData<PopupManager> = _popupManager
+
+    private val _isPopupInEditMode = MutableLiveData(false)
+    // Depicts whether the PopupManager is currently in editing mode
+    // When in edit mode the user can edit the values of fields of the Popup
+    val isPopupInEditMode: LiveData<Boolean> = _isPopupInEditMode
+
+    private val _showSavePopupErrorEvent = MutableLiveData<Event<String>>()
+    // This event is raised when an error is encountered while trying to save edits on a popup.
+    // It passes the exception message to all the observers in the observeEvent() method.
+    val showSavePopupErrorEvent: LiveData<Event<String>> = _showSavePopupErrorEvent
+
+    private val _showDeletePopupErrorEvent = MutableLiveData<Event<String>>()
+    // This event is raised when an error is encountered while trying to delete a popup.
+    // It passes the exception message to all the observers in the observeEvent() method.
+    val showDeletePopupErrorEvent: LiveData<Event<String>> = _showDeletePopupErrorEvent
+
+    private val _showSavingProgressEvent = MutableLiveData<Event<Boolean>>()
+    // This event is raised when the save operation begins/ends.
+    // It passes true to all the observers in the observeEvent() method when the save operation
+    // is initiated and false when it ends.
+    val showSavingProgressEvent: LiveData<Event<Boolean>> = _showSavingProgressEvent
+
+    private val _confirmCancelPopupEditingEvent = MutableLiveData<Event<Unit>>()
+    // This event is raised when the user cancels the edit mode on the Popup.
+    // It is used for showing confirmation dialog to the user, before calling cancelEditing()
+    val confirmCancelPopupEditingEvent: LiveData<Event<Unit>> = _confirmCancelPopupEditingEvent
+
+    private val _dismissPopupEvent = MutableLiveData<Event<Unit>>()
+    // This event is raised when the user taps on the close button to dismiss the popup.
+    val dismissPopupEvent: LiveData<Event<Unit>> = _dismissPopupEvent
+
+    private val _confirmDeletePopupEvent = MutableLiveData<Event<Unit>>()
+    // This event is raised when the user presses the button to delete the Popup.
+    // It is used for showing confirmation dialog to the user, before calling deletePopup()
+    val confirmDeletePopupEvent: LiveData<Event<Unit>> = _confirmDeletePopupEvent
+
+    /**
+     * Updates popup property to set the popup field values being displayed in
+     * the bottom sheet. Creates PopupManager for PopupView to perform edit operations.
+     *
+     * @param popup
+     */
+    fun setPopup(popup: Popup) {
+        _popup.value = popup
+        _popupManager.value = PopupManager(getApplication(), _popup.value)
+    }
+
+    /**
+     * Enables/disables edit mode on the PopupView
```

Comment:
Review all doc for missing full stops. Will only mention it here...
```suggestion
     * Enables/disables edit mode on the PopupView.
```

Reply:
updated.